### ISSUE TITLE
Suggestion: System.Math Performance improvement

### ISSUE THREAD

**Issue by dangi12012** [dotnet/corefx#29069]
In Performance Critical code System.Math is lacking following operators. (Natively supported by Gcc and Visual studios C++ compiler but slow in C#)

**Fused Multiply Add**
```
x * y + z //Not natively compiled by compiler to single Instruction
Math.Fadd(x,y,z) //Math operator for that
```

**Fused Division and Modulus with ValueTuple** (Handy for 1D array with 2D indices) updated for Valuetuple which is faster than out + return.
```
//p.x = index / 3;
//p.y = index % 3;

var (x, y) = Math.DivRem(index,3)
//ValueTuple<int, int>DivMod(int a, int b)
```

**Support for Float Square Root** (Very Handy and badly needed in 3d apps where we dont want to cast to double and back for a single sqrt)

`float root = Math.Sqrt(float nr)`

**Comment by tannergooding** [comment]
We are exposing (netcoreapp2.0 or higher) `float` overloads for the entire `System.Math` library via `System.MathF` (https://source.dot.net/#System.Private.CoreLib/shared/System/MathF.cs,2c9a628505489b33,references and https://source.dot.net/#System.Private.CoreLib/src/System/MathF.CoreCLR.cs,2c9a628505489b33).

For `Fused Multiply-Add`, it will be exposed (on a per CPU platform basis) via the Hardware Intrinsics feature. The issue with exposing a "general-purpose" API is that it is not guaranteed to be supported on every platform. (I think I have an old proposal sitting around for this as well, for the last time this went to API review/was discussed).

**Comment by benaadams** [comment]
Fused Multiply Add probably should be a pattern recognized by the Jit rather than an api?

Its a bit of a 🔥 on x86 https://en.wikipedia.org/wiki/FMA_instruction_set

> The incompatibility between Intel's FMA3 and AMD's FMA4 is due to both companies changing plans without coordinating coding details with each other. AMD changed their plans from FMA3 to FMA4 while Intel changed their plans from FMA4 to FMA3 almost at the same time. The history can be summarized as follows:
>
> * August 2007: AMD announces the SSE5 instruction set, which includes 3-operand FMA instructions. A new coding scheme (DREX) is introduced for allowing instructions to have three operands.[11]
> * April 2008: Intel announces their AVX and FMA instruction sets, including 4-operand FMA instructions. The coding of these instructions uses the new VEX coding scheme,[12] which is more flexible than AMD's DREX scheme.
> * December 2008: Intel changes the specification for their FMA instructions from 4-operand to 3-operand instructions. The VEX coding scheme is still used.[13]
> * May 2009: AMD changes the specification of their FMA instructions from the 3-operand DREX form to the 4-operand VEX form, compatible with the April 2008 Intel specification rather than the December 2008 Intel specification.[14]
> * October 2011: AMD Bulldozer processor supports FMA4.[15]
> * January 2012: AMD announces FMA3 support in future processors codenamed Trinity and Vishera; they are based on the Piledriver architecture.[16]
> * May 2012: AMD Piledriver processor supports both FMA3 and FMA4.[15]
> * June 2013: Intel Haswell processor supports FMA3.[17]
> * February 2017 The first generation of AMD Ryzen processors officially supports FMA3, but not FMA4 according to the CPUID instruction.[18] There has been confusion regarding whether FMA4 was implemented or not on this processor due to errata in the initial patch to the GNU Binutils package that has since been rectified.[19][20] While the FMA4 instructions seem to work according to some tests, they can also give wrong results.[7] Additionally, the initial Ryzen CPUs could be crashed by a particular sequence of FMA3 instructions. It has since been resolved by an updated CPU microcode.[21]

**Comment by tannergooding** [comment]
I wouldn't say its a 🔥 at all. There was initially some differences but AMD ultimately supports FMA3 and is only supporting FMA3 in the future (at least based on my current understanding, etc).

dotnet/coreclr#17541 is basically a dupe of dotnet/coreclr#12090 (the I mentioned I logged above, but I didn't go searching for 😄). I commented [here](https://github.com/dotnet/coreclr/issues/17541#issuecomment-381138004) to give a summary of the issues with recognizing FMA patterns natively and why they (probably) need to behind some kind of switch.
